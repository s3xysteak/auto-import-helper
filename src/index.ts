import type { PathLike } from 'node:fs'
import * as fs from 'node:fs/promises'

export type ImportParam<T extends PropertyKey> = Partial<Record<T, string>>
export type ImportReturns = Record<string, Array<string | [string, string]>>

/**
 * ## example
 * ```js
 * // import.js
 * export default createImport('lodash', ['map', 'filter'])
 *
 * // user
 * import LodashImport from './import.js'
 * export default defineConfig({
 *   AutoImport({
 *     imports: [
 *       LodashImport()
 *     ]
 *   })
 * })
 * ```
 */
export function createImport<T extends string[]>(packageName: string, exports: T) {
  return (map?: ImportParam<typeof exports[number]>): ImportReturns => {
    return {
      [packageName]: exports.map(v => map && (map as any)[v] ? [v, (map as any)[v]] : v),
    }
  }
}

/**
 * ## example
 * ```js
 * // resolver.js
 * export default createResolver('lodash', ['map', 'filter'])
 *
 * // user
 * import LodashResolver from './resolver.js'
 * export default defineConfig({
 *   AutoImport({
 *     resolvers: [
 *       LodashResolver()
 *     ]
 *   })
 * })
 * ```
 */
export function createResolver<T extends string[]>(packageName: string, exports: T) {
  return (map?: ImportParam<typeof exports[number]>) =>
    (name: string) => {
      if (!exports.includes(name))
        return

      return {
        name,
        from: packageName,
        ...map && (map as any)[name] ? { as: (map as any)[name] } : {},
      }
    }
}

/**
 * Replace the first array after auto-import-helper comment by a target string
 * @param str A string which will be transformed
 * @param insert If array, stringify it by `JSON.stringify`
 *
 * ## example
 *
 * /&ast;&ast; auto-import-helper &ast;/
 *
 * const foo = []
 *
 * ```
 * transformString(aboveString, ['bar']) // ->
 * ```
 *
 * /&ast;&ast; auto-import-helper &ast;/
 *
 * const foo = ['bar']
 */
export function transformString(str: string, insert: string | string[]) {
  return str.replace(
    /(\/\*{1,2}\s*auto-import-helper\s*\*\/[^[]*)(\[.*?\])/s,
    (_, $1) => `${$1}${Array.isArray(insert) ? JSON.stringify(insert) : insert}`,
  )
}

/**
 * Make `transformString` work on a js file.
 *
 * ## example
 * ```js
 * await transformFile(
 *   './src/import.ts',
 *   Object.keys(await import('./src/index'))
 * )
 * ```
 */
export async function transformFile(path: PathLike | fs.FileHandle, exports: string[]) {
  const content = await fs.readFile(path, { encoding: 'utf-8' })
  await fs.writeFile(path, transformString(content, exports))
}
