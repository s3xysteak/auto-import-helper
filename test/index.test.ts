import * as fs from 'node:fs/promises'
import { expect, it } from 'vitest'

import { createImport, createResolver, transformFile, transformString } from '../src'
import * as labTs from './lab/lab-ts'

it('import', () => {
  const v = createImport('lodash', ['map', 'filter'] as const)

  expect(v()).toEqual({ lodash: ['map', 'filter'] })
  expect(v({ map: 'myMap' })).toEqual({ lodash: [['map', 'myMap'], 'filter'] })
})

it('resolver', () => {
  const v = createResolver('lodash', ['map', 'filter'] as const)
  const fn = v({ map: 'myMap' })

  expect(fn('a')).toBeUndefined()
  expect(fn('map')).toEqual({ name: 'map', as: 'myMap', from: 'lodash' })
  expect(fn('filter')).toEqual({ name: 'filter', from: 'lodash' })
})

it('module', () => {
  const v = createImport('lab-ts', Object.keys(labTs))
  expect(v()).toEqual({ 'lab-ts': ['oneType', 'getTwoType'] })
})

it('transformString', () => {
  const content = `
/** auto-import-helper */
const foo = []
`
  expect(transformString(content, ['bar'])).toMatchSnapshot()
})

it('transformFile', async () => {
  const path = new URL('./lab/import.ts', import.meta.url)
  const init = () => fs.writeFile(path, `import { createImport } from '../../src'

/** auto-import-helper */
export default createImport('name', [])
`)
  // init
  await init()

  await transformFile(path, Object.keys(await import(new URL('./lab/lab-ts.ts', import.meta.url).href)))
  expect(await fs.readFile(path, 'utf-8')).toMatchSnapshot()

  await init()
})
