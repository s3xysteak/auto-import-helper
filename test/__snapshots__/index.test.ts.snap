// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`transformFile 1`] = `
"import { createImport } from '../../src'

/** auto-import-helper */
export default createImport('name', ["oneType","getTwoType"])
"
`;

exports[`transformString 1`] = `
"
/** auto-import-helper */
const foo = ["bar"]
"
`;
